pkg/destroy/azure/azure.go

```diff
@@ -0,0 +1,225 @@
+package azure
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/preview/dns/mgmt/2018-03-01-preview/dns"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	session "github.com/openshift/installer/pkg/asset/installconfig/azure"
+	"github.com/openshift/installer/pkg/destroy"
+	"github.com/openshift/installer/pkg/types"
+)
+
+// ClusterUninstaller holds the various options for the cluster we want to delete.
+type ClusterUninstaller struct {
+	SubscriptionID string
+	Authorizer     autorest.Authorizer
+
+	InfraID string
+
+	Logger logrus.FieldLogger
+
+	resourceGroupsClient resources.GroupsGroupClient
+	zonesClient          dns.ZonesClient
+	recordsClient        dns.RecordSetsClient
+}
+
+func (o *ClusterUninstaller) configureClients() {
+	resourceGroupsClient := resources.NewGroupsGroupClient(o.SubscriptionID)
+	resourceGroupsClient.Authorizer = o.Authorizer
+	o.resourceGroupsClient = resourceGroupsClient
+
+	zonesClient := dns.NewZonesClient(o.SubscriptionID)
+	zonesClient.Authorizer = o.Authorizer
+	o.zonesClient = zonesClient
+
+	recordsClient := dns.NewRecordSetsClient(o.SubscriptionID)
+	recordsClient.Authorizer = o.Authorizer
+	o.recordsClient = recordsClient
+}
+
+// Run is the entrypoint to start the uninstall process.
+func (o *ClusterUninstaller) Run() error {
+	o.configureClients()
+	group := o.InfraID + "-rg"
+	o.Logger.Debug("deleting public records")
+	if err := deletePublicRecords(context.TODO(), o.zonesClient, o.recordsClient, o.Logger, group); err != nil {
+		o.Logger.Debug(err)
+		return errors.Wrap(err, "failed to delete public DNS records")
```

Comment:
This will fail instead of endlessly retrying (which is what the AWS destroyer does now).  Sometimes (e.g. stuck resource), failing is good.  Sometimes (e.g. network hiccup), retrying is good.  Do we have a design preference?  Or is this behavior going to be a per-platform choice?

Reply:
Azure only makes 2 calls, delete public records and another is delete resource group. Also the underlying impl has retries for `retryable` errors..

So I think this is good start. and We can make it retryable on our end when we think that will become necessary.